### ISSUE TITLE
UI: Scientific notation 'ts' events are 1e15us in the past, compared to Decimal notation 'ts' events

### ISSUE THREAD

**Issue by mw-cwilcox** [google/perfetto#558]
The following Google Trace Event JSON input describes two different parent-child pairs of events, where the parent events start 6us apart.  The first parent-child 'ts' fields are formatted as scientific notation, and the second parent-child 'ts' fields are formatted as decimal notation.  Perfetto shows these events as starting 1e15us (one quadrillion microseconds) apart.

```json
[
    {
        "cat":"perf",
        "dur":4.75,
        "name":"parentEvent",
        "ph":"X",
        "pid":3933029,
        "tid":1,
        "ts":1.6921085481321545e+15
    },
    {
        "cat":"perf",
        "dur":2.0,
        "name":"childEvent",
        "ph":"X",
        "pid":3933029,
        "tid":1,
        "ts":1.6921085481321572e+15
    },
    {
        "cat":"perf",
        "dur":4.5,
        "name":"parentEvent",
        "ph":"X",
        "pid":3933029,
        "tid":1,
        "ts":1692108548132160.5
    },
    {
        "cat":"perf",
        "dur":2.5,
        "name":"childEvent",
        "ph":"X",
        "pid":3933029,
        "tid":1,
        "ts":1692108548132162.5
    }
]
```

![Screenshot_2023-08-17_14-26-08 - Perfetto Sci Dec](https://github.com/google/perfetto/assets/44209640/d5d8b936-e668-4a0a-b924-d53e871b68ed)

(Image produced with version v35.0-8019b8809, 'Stable'.  'Canary' produced similar results.)

(Speculation: This may relate to Issue #426)

**Comment by LalitMaganti** [comment]
Our response to this issue falls squarely under our published policy for supporting features in JSON traces: [policy here](https://perfetto.dev/docs/faq#why-does-perfetto-not-support-lt-some-obscure-json-format-feature-gt-).

TLDR: scientific notation in the JSON trace format is quite an obscure feature and we will not go out of our way to add support for this. Patches for fixing this our welcome.

**Comment by mw-cwilcox** [comment]
To clarify:

- the Chrome [Trace Event Format](https://docs.google.com/document/d/1CvAClvFfyA5R-PhYUmn5OOQtYMH4h6I0nSsKchNAySU/preview) implies that a 'ts' field is a JSON Number
- a [JSON](https://www.ecma-international.org/wp-content/uploads/ECMA-404_2nd_edition_december_2017.pdf) Number is "a sequence of decimal digits with no superfluous leading zero. It may have a preceding minus
sign (U+002D). It may have a fractional part prefixed by a decimal point (U+002E). **It may have an exponent**,
prefixed by e (U+0065) or E (U+0045) and optionally + (U+002B) or – (U+002D).  The digits are the code
points U+0030 through U+0039."
- you said "scientific notation in the JSON trace format is quite an obscure feature"
- `JSON.parse` works correctly on above input on the browser I'm using to run Perfetto.

So, does Perfetto use a different JSON parser than the `JSON.parse` one available to all JS/TS programs, and that different parser is feature incomplete?

**Comment by LalitMaganti** [comment]
All trace parsing (including JSON parsing) is performed in C++ so does not use `JSON.parse` at all. Specifically the timestamp coercion logic is [here](https://cs.android.com/android/platform/superproject/main/+/main:external/perfetto/src/trace_processor/importers/json/json_utils.cc;l=62;drc=ee8948464aaf80b7bdaf666ff496b76a3833c33a).

**Comment by LalitMaganti** [comment]
Should be fixed by https://r.android.com/2718613. Thanks for the contribution here @mw-cwilcox!
